pkg/monitor/monitor/monitor_manager.go

```diff
@@ -233,29 +233,39 @@ func (m *MonitorManager) syncTidbMonitorStatefulset(tc *v1alpha1.TidbCluster, dc
 		klog.Infof("Wait for the smooth migration to be done successfully for tm [%s/%s]", ns, name)
 		return nil
 	}
-	newMonitorSts, err := getMonitorStatefulSet(sa, cm, secret, monitor, tc, dc)
-	if err != nil {
-		klog.Errorf("Fail to generate statefulset for tm [%s/%s], err: %v", ns, name, err)
-		return err
-	}
-
-	oldMonitorSetTmp, err := m.deps.StatefulSetLister.StatefulSets(ns).Get(GetMonitorObjectName(monitor))
-	if err != nil && !errors.IsNotFound(err) {
-		return fmt.Errorf("syncTidbMonitorStatefulset: fail to get sts %s for cluster %s/%s, error: %s", GetMonitorObjectName(monitor), ns, name, err)
+	shards := int32(1)
+	if monitor.Spec.Shards != nil && *monitor.Spec.Shards > 1 {
+		shards = *monitor.Spec.Shards
 	}
-	setNotExist := errors.IsNotFound(err)
-	if setNotExist {
-		err = member.SetStatefulSetLastAppliedConfigAnnotation(newMonitorSts)
+	for shard := int32(0); shard < shards; shard++ {
+		newMonitorSts, err := getMonitorStatefulSet(sa, cm, secret, monitor, tc, dc, shard)
 		if err != nil {
+			klog.Errorf("Fail to generate statefulset for tm [%s/%s], err: %v", ns, name, err)
 			return err
 		}
-		if err := m.deps.StatefulSetControl.CreateStatefulSet(monitor, newMonitorSts); err != nil {
+		stsName := GetMonitorShardName(monitor, shard)
+		oldMonitorSetTmp, err := m.deps.StatefulSetLister.StatefulSets(ns).Get(stsName)
+		if err != nil && !errors.IsNotFound(err) {
+			return fmt.Errorf("syncTidbMonitorStatefulset: fail to get sts %s for cluster %s/%s, error: %s", stsName, ns, name, err)
+		}
+		setNotExist := errors.IsNotFound(err)
+		if setNotExist {
+			err = member.SetStatefulSetLastAppliedConfigAnnotation(newMonitorSts)
+			if err != nil {
+				return err
+			}
+			if err := m.deps.StatefulSetControl.CreateStatefulSet(monitor, newMonitorSts); err != nil {
+				return err
+			}
+			return controller.RequeueErrorf("TidbMonitor: [%s/%s], waiting for tidbmonitor running", ns, name)
```

Comment:
Separate Create  and Update operation?

Reply:
er... or can we create and update all shards in one sync loop and returning:
- errors from `member.UpdateStatefulSet` if there is any
- `RequeueErrorf` if any create called

?

Reply:
Is it okay to return when an error is encountered in a sync loop?

Reply:
ok

Reply:
@mikechengwei I think @csuzhangxc 's first comment https://github.com/pingcap/tidb-operator/pull/3990/files?file-filters%5B%5D=.go#r638638558 should be addressed, if you return here, it will return after each creation and the sts creation will go through several reconcile loops.

Reply:
Ok. I have optimized . I create all sts in one loop.